src/graph_gui.cpp

```diff
@@ -1506,6 +1524,269 @@ struct PerformanceRatingDetailWindow : Window {
+			int rect_x = clk_dif + (rtl ? r.right - 12 : r.left + WD_FRAMERECT_LEFT);
+
+			GfxFillRect(rect_x, y + clk_dif, rect_x + 8, y + 5 + clk_dif, PC_BLACK);
+			GfxFillRect(rect_x + 1, y + 1 + clk_dif, rect_x + 7, y + 4 + clk_dif, cs->legend_colour);
+			SetDParam(0, cs->name);
+			DrawString(rtl ? r.left : x + 14 + clk_dif, (rtl ? r.right - 14 + clk_dif : r.right), y + clk_dif, STR_GRAPH_CARGO_PAYMENT_CARGO);
+
+			y += this->line_height;
+		}
+	}
+
+	void OnClick(Point pt, int widget, int click_count) override
+	{
+		switch (widget) {
+			case WID_IP_ENABLE_CARGOES:
+				/* Remove all cargoes from the excluded lists. */
+				_legend_excluded_cargo = 0;
+				this->excluded_data = 0;
+				this->SetDirty();
+				break;
+
+			case WID_IP_DISABLE_CARGOES: {
+				/* Add all cargoes to the excluded lists. */
+				int index = 0;
+				const Industry *i = Industry::Get(this->window_number);
+				const CargoSpec *cs;
+				for (size_t j = 0; j < INDUSTRY_NUM_OUTPUTS; j++) {
+					if (!IsCargoIDValid(i->produced_cargo[j])) continue;
+
+					cs = CargoSpec::Get(i->produced_cargo[j]);
+
+					SetBit(_legend_excluded_cargo, cs->Index());
+					SetBit(this->excluded_data, index);
+					index++;
+				}
+				this->SetDirty();
+				break;
+			}
+
+			case WID_IP_MATRIX: {
+				uint row = this->vscroll->GetScrolledRowFromWidget(pt.y, this, WID_IP_MATRIX, 0, this->line_height);
+				if (row >= this->vscroll->GetCount()) return;
+
+				const Industry *i = Industry::Get(this->window_number);
+				const CargoSpec *cs;
+				for (size_t j = 0; j < INDUSTRY_NUM_OUTPUTS; j++) {
+					if (!IsCargoIDValid(i->produced_cargo[j])) continue;
+					if (row-- > 0) continue;
+
+					cs = CargoSpec::Get(i->produced_cargo[j]);
+
+					ToggleBit(_legend_excluded_cargo, cs->Index());
+					this->UpdateExcludedData();
+					this->SetDirty();
+					break;
+				}
+				break;
+			}
+		}
+	}
+
+	void SetStringParameters(int widget) const override
+	{
+		if (widget == WID_IP_CAPTION) SetDParam(0, this->window_number);
+	}
+
+	void OnResize() override
+	{
+		this->vscroll->SetCapacityFromWidget(this, WID_IP_MATRIX);
+	}
+
+	void UpdateStatistics(bool initialize) override
+	{
+		CargoTypes excluded_cargo = this->excluded_data;
+		this->UpdateExcludedData();
+
+		int mo = _cur_month - this->num_vert_lines;
+		int yr = _cur_year;
+		while (mo < 0) {
+			yr--;
+			mo += 12;
+		}
+
+		if (!initialize && this->excluded_data == excluded_cargo && this->num_on_x_axis == this->num_vert_lines &&
+				this->year == yr && this->month == mo) {
+			/* There's no reason to get new stats */
+			return;
+		}
+
+		this->year = yr;
+		this->month = mo;
+
+		int index = 0;
+		const Industry *i = Industry::Get(this->window_number);
+		const CargoSpec *cs;
+		for (size_t k = 0; k < INDUSTRY_NUM_OUTPUTS; k++) {
+			if (!IsCargoIDValid(i->produced_cargo[k])) continue;
+
+			cs = CargoSpec::Get(i->produced_cargo[k]);
+
+			this->colours[index] = cs->legend_colour;
+			for (uint j = 0; j < 24; j++) {
```

Comment:
magic 24 in a few places should be replaced with a constant. Presumably it's 2 years, but...